oak_sev_guest/src/ghcb.rs

```diff
@@ -318,6 +328,56 @@ where
         )?)
     }
 
+    /// Read a 32-bit value from a MMIO memory address via the MMIO Access protocol.
+    ///
+    /// See Section 4.1.5 in <https://www.amd.com/content/dam/amd/en/documents/epyc-technical-docs/specifications/56421-guest-hypervisor-communication-block-standardization.pdf>.
+    pub fn mmio_read_u32(&mut self, source: PhysAddr) -> Result<u32, &'static str> {
+        let gpa_base = self.get_gpa().as_u64();
+        let ghcb = self.ghcb.as_mut();
+
+        ghcb.reset();
+        ghcb.sw_exit_code = SW_EXIT_CODE_MMIO_READ;
+        ghcb.sw_exit_info_1 = source.as_u64();
+        ghcb.sw_exit_info_2 = core::mem::size_of::<u32>() as u64;
+        // Use the shared_buffer as the unencrypted guest memory. This is mandated as of Version
+        // 2 of the communication block.
+        ghcb.sw_scratch = gpa_base + 0x800;
+        ghcb.valid_bitmap = BASE_VALID_BITMAP | ValidBitmap::SW_SCRATCH;
+
+        self.do_vmg_exit()?;
+        // Safety: all possible values are valid for u32.
+        Ok(unsafe { core::mem::transmute_copy(&self.ghcb.as_ref().shared_buffer) })
```

Comment:
optional: the unsafe block can probably be removed by using as_bytes_mut (I think both AsBytes and FromBytes are implemented on u32) and as_bytes below.

Reply:
Good point, I don't need all this unsafeness with plain u32s.